src/ui/server/controllers/insurance/policy-and-export/single-contract-policy/validation/rules/contract-completion-date.ts

```diff
@@ -0,0 +1,122 @@
+import { FIELD_IDS, PRODUCT } from '../../../../../../constants';
+import { add, isAfter, isBefore, isPast, isSameDay } from 'date-fns';
+import { ERROR_MESSAGES } from '../../../../../../content-strings';
+import generateValidationErrors from '../../../../../../helpers/validation';
+import { objectHasProperty } from '../../../../../../helpers/object';
+import createTimestampFromNumbers from '../../../../../../helpers/date/create-timestamp-from-numbers';
+import { isNumber } from '../../../../../../helpers/number';
+import { RequestBody } from '../../../../../../../types';
+
+const {
+  INSURANCE: {
+    POLICY_AND_EXPORTS: {
+      CONTRACT_POLICY: {
+        REQUESTED_START_DATE,
+        SINGLE: { CONTRACT_COMPLETION_DATE },
+      },
+    },
+  },
+} = FIELD_IDS;
+
+const {
+  INSURANCE: {
+    POLICY_AND_EXPORTS: {
+      CONTRACT_POLICY: {
+        SINGLE: { [CONTRACT_COMPLETION_DATE]: ERROR_MESSAGE },
+      },
+    },
+  },
+} = ERROR_MESSAGES;
+
+export const MAXIMUM_FUTURE_YEARS = 2;
+
+const DATE_INPUT_IDS = {
+  start: {
+    day: `${REQUESTED_START_DATE}-day`,
+    month: `${REQUESTED_START_DATE}-month`,
+    year: `${REQUESTED_START_DATE}-year`,
+  },
+  end: {
+    day: `${CONTRACT_COMPLETION_DATE}-day`,
+    month: `${CONTRACT_COMPLETION_DATE}-month`,
+    year: `${CONTRACT_COMPLETION_DATE}-year`,
+  },
+};
+
+const getDateInputValues = (formBody: RequestBody) => ({
+  start: {
+    day: Number(formBody[`${REQUESTED_START_DATE}-day`]),
+    month: Number(formBody[`${REQUESTED_START_DATE}-month`]),
+    year: Number(formBody[`${REQUESTED_START_DATE}-year`]),
+  },
+  end: {
+    day: Number(formBody[`${CONTRACT_COMPLETION_DATE}-day`]),
+    month: Number(formBody[`${CONTRACT_COMPLETION_DATE}-month`]),
+    year: Number(formBody[`${CONTRACT_COMPLETION_DATE}-year`]),
+  },
+});
+
+/**
+ * contractCompletionDateRules
+ * Check submitted form data for errors with the requested start date field
+ * Returns generateValidationErrors if there are any errors.
+ * @param {Express.Response.body} Express response body
+ * @param {Object} Errors object from previous validation errors
+ * @returns {Object} Validation errors
+ */
+const contractCompletionDateRules = (formBody: RequestBody, errors: object) => {
+  const updatedErrors = errors;
+  let errorMessage;
+
+  const inputValues = getDateInputValues(formBody);
```

Comment:
Do we need void `inputValues` check and/or ensure essential properties exists before proceeding further `inputValues.end.day`?

Reply:
I don't believe so. `getDateInputValues` will always return the object structure with `start.day` and `end.day` etc. If the values do not exist in the form body, the object structure will still be generated and the validation will still work.